nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/replication/StandardUploadRequestReplicator.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    private <T> Future<T> performUploadAsync(final NodeIdentifier nodeId, final UploadRequest<T> uploadRequest, final File contents) {
+        final CompletableFuture<T> future = new CompletableFuture<>();
+        Thread.ofVirtual().name("Replicate upload to " + nodeId.getApiAddress()).start(() -> {
+            try {
+                final T response = replicateRequest(nodeId, uploadRequest, contents);
+                logger.debug("Successfully replicated upload request for {} to {}", uploadRequest.getFilename(), nodeId.getApiAddress());
+                future.complete(response);
+            } catch (final IOException | UploadRequestReplicationException e) {
+                future.completeExceptionally(e);
+            }
+        });
+
+        return future;
+    }
+
+    private <T> T replicateRequest(final NodeIdentifier nodeId, final UploadRequest<T> uploadRequest, final File contents) throws IOException {
+        final URI exampleRequestUri = uploadRequest.getExampleRequestUri();
+
+        final URI requestUri = new StandardHttpUriBuilder()
+                .scheme(exampleRequestUri.getScheme())
+                .host(nodeId.getApiAddress())
+                .port(nodeId.getApiPort())
+                .encodedPath(exampleRequestUri.getPath())
+                .build();
+
+        final NiFiUser user = uploadRequest.getUser();
+        final String filename = uploadRequest.getFilename();
+
+        try (final InputStream inputStream = new FileInputStream(contents)) {
+            final HttpRequestBodySpec request = webClientService.post()
+                    .uri(requestUri)
+                    .body(inputStream, OptionalLong.of(inputStream.available()))
+                    .header(CONTENT_TYPE_HEADER, UPLOAD_CONTENT_TYPE)
+                    .header(FILENAME_HEADER, filename)
+                    // Special NiFi-specific headers to indicate that the request should be performed and not replicated to the nodes
+                    .header(RequestReplicator.REQUEST_EXECUTION_HTTP_HEADER, "true")
+                    .header(RequestReplicator.REPLICATION_INDICATOR_HEADER, "true")
+                    .header(ProxiedEntitiesUtils.PROXY_ENTITIES_CHAIN, ProxiedEntitiesUtils.buildProxiedEntitiesChainString(user))
+                    .header(ProxiedEntitiesUtils.PROXY_ENTITY_GROUPS, ProxiedEntitiesUtils.buildProxiedEntityGroupsString(user.getIdentityProviderGroups()));
+
+            logger.debug("Replicating upload request for {} to {}", filename, nodeId);
+
+            try (final HttpResponseEntity response = request.retrieve()) {
+                final int statusCode = response.statusCode();
+                if (!HttpResponseStatus.isSuccessful(statusCode)) {
+                    final String responseMessage = IOUtils.toString(response.body(), StandardCharsets.UTF_8);
+                    throw new UploadRequestReplicationException("Failed to replicate upload request to " + nodeId + " - " + responseMessage, statusCode);
```

Comment:
Recommend minor adjustment to message formatting:
```suggestion
                    throw new UploadRequestReplicationException("Failed to replicate upload request to [%s] %s".formatted(nodeId, responseMessage), statusCode);
```